Review of Azure Storage Queues multi-accounts (#7885)
Particular/docs.particular.net: transports/azure-storage-queues/multi-storageaccount-support.md

![Scale out with multiple storage accounts](azure03.png "width=500")

To determine whether your system may benefit from scaling out to multiple storage accounts, refer to the the Scale targets table in the Azure [Scalability and performance targets for Queue Storage](https://learn.microsoft.com/en-us/azure/storage/queues/scalability-targets) article, which define when throttling starts to occur.

For additional guidance on considerations when developing a system using Azure Storage Queues, see the article on [Performance and scalability checklist for Queue Storage](https://learn.microsoft.com/en-us/azure/storage/queues/storage-performance-checklist).

> [!NOTE]
> Use real Azure storage accounts. Do not use Azure storage emulator as it only supports a single fixed account named "devstoreaccount1".

> [!NOTE]

> > There are limits to how much using multiple storage accounts increase throughput. Consider using [scale units as a comprehensive scaling strategy](https://learn.microsoft.com/en-us/azure/well-architected/performance-efficiency/scale-partition#choose-a-scaling-strategy) to address higher throughput and reliability needs.

Author: ```suggestion
> There are limits to how much increasing storage accounts increases throughput. Consider using [scale units as a comprehensive scaling strategy](https://learn.microsoft.com/en-us/azure/well-architected/performance-efficiency/scale-partition#choose-a-scaling-strategy) to address higher throughput and reliability needs.
```

## NServiceBus routing with multiple storage accounts

The preferred way to route when using multiple accounts is to register endpoints with their associated storage accounts. This should be done using connection string aliases   instead of raw connection strings.

### Connection string aliases

> Using aliases for connection strings prevents exposing sensitive data that may be in connection strings. An empty string is the default alias which works when using a single account. When using multiple accounts, distinct aliases have to be registered for each storage account.

Reviewer: Can this be simplified further?

Mentioning the empty string seems like we are leaking our internal implementation.

I tried to compress this a bit further.

```suggestion
Using distinct aliases for each storage account connection string prevents exposing sensitive data and is required when using multiple accounts.
```

To enable sending from `account_A` to `account_B`, the following configuration needs to be applied in the `account_A` endpoint:

snippet: AzureStorageQueueUseMultipleAccountAliasesInsteadOfConnectionStrings1

Aliases can be provided for both the endpoint's connection strings as well as other accounts' connection strings. This enables using the `@` notation for destination addresses like `queue_name@accountAlias`.

> snippet: storage_account_routing_send_options_alias

Author: I think I addressed this?

> [!NOTE]
> The examples above use different values for the default account aliases. Using the same name, such as `default`, to represent different storage accounts in different endpoints is highly discouraged as it introduces ambiguity in resolving addresses like `queue@default` and may cause issues when e.g. replying. In that case an address is interpreted as a reply address, the name `default` will point to a different connection string.

> [!NOTE]
> This feature is currently NOT compatible with ServiceControl. A [ServiceControl transport adapter](/servicecontrol/transport-adapter.md) is required in order to leverage both.

### Using registered endpoints

In order to route a message to an endpoint without having to specify the destination each time, the endpoint can be registered for a given command type, assembly, or namespace.

snippet: storage_account_routing_registered_endpoint

Once the endpoint is registered, no send options need to be specified.

snippet: storage_account_routing_send_registered_endpoint

> 

Author: Ditto https://github.com/Particular/docs.particular.net/pull/7885/files#r2510450483

It is common for systems running on Azure Storage Queues to depend on a single storage account. However, there is a potential for throttling issues once the maximum number of concurrent requests to the storage account is reached. Multiple storage accounts can be used to overcome this. 

![Scale out with multiple storage accounts](azure03.png "width=500")

To determine whether your system may benefit from scaling out to multiple storage accounts, refer to the the Scale targets table in the Azure [Scalability and performance targets for Queue Storage](https://learn.microsoft.com/en-us/azure/storage/queues/scalability-targets) article, which define when throttling starts to occur.

For additional guidance on considerations when developing a system using Azure Storage Queues, see the article on [Performance and scalability checklist for Queue Storage](https://learn.microsoft.com/en-us/azure/storage/queues/storage-performance-checklist).

> [!NOTE]

> > Use real Azure storage accounts. Do not use Azure storage emulator as it only supports a single fixed account named devstoreaccount1.".

Author: ```suggestion
> Use real Azure storage accounts. Do not use Azure storage emulator as it only supports a single fixed account named "devstoreaccount1".
```

> [!NOTE]
> There are limits to how much using multiple storage accounts increase throughput. Consider using [scale units as a comprehensive scaling strategy](https://learn.microsoft.com/en-us/azure/well-architected/performance-efficiency/scale-partition#choose-a-scaling-strategy) to address higher throughput and reliability needs.

## NServiceBus routing with multiple storage accounts

> The preferred way to route when using multiple accounts is to register endpoints with their associated storage accounts.

Author: ```suggestion
The preferred way to route when using multiple accounts is to register endpoints with their associated storage accounts. This should be done using connection string aliases   instead of raw connection strings.
```
